plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/CabalAdd.hs

```diff
@@ -0,0 +1,185 @@
+import           System.Directory                              (doesFileExist,
+                                                                listDirectory)
+
+import           Data.Aeson.Types                              (FromJSON,
+                                                                ToJSON, toJSON)
+import           Data.ByteString                               (ByteString)
+import qualified Data.ByteString.Char8                         as B
+import           Data.List.NonEmpty                            (NonEmpty (..),
+                                                                fromList)
+import           Data.Maybe                                    (fromJust)
+import           Distribution.Client.Add                       as Add
+import           Distribution.Compat.Prelude                   (Generic)
+import           Distribution.PackageDescription               (packageDescription,
+                                                                specVersion, GenericPackageDescription)
+import           Distribution.PackageDescription.Configuration (flattenPackageDescription)
+import           Distribution.Pretty                           (pretty)
+import           Distribution.Simple.BuildTarget               (BuildTarget,
+                                                                buildTargetComponentName,
+                                                                readBuildTargets)
+import           Distribution.Verbosity                        (silent,
+                                                                verboseNoStderr)
+import           System.FilePath                               (dropFileName,
+                                                                makeRelative,
+                                                                splitPath,
+                                                                takeExtension,
+                                                                (</>))
+import           System.IO.Unsafe                              (unsafeInterleaveIO)
+import           Text.PrettyPrint                              (render)
+import           Text.Regex.TDFA
+
+-- | Given a path to a haskell file, finds all cabal files paths
+--   sorted from the closest to the farthest.
+--   Gives all found paths all the way to the root directory.
+findResponsibleCabalFile :: FilePath -> IO [FilePath]
+findResponsibleCabalFile haskellFilePath = do
+  contents <- mapM (unsafeInterleaveIO . listDirectory) allDirPaths
+  let objectWithPaths = concat $ zipWith (\path content -> map (path </>) content) allDirPaths contents
+  let objectCabalExtension = filter (\c -> takeExtension c == ".cabal") objectWithPaths
+  cabalFiles <- filterM (\c -> doesFileExist c) objectCabalExtension
+  pure $ reverse cabalFiles -- sorted from closest to the haskellFilePath
+  where dirPath = dropFileName haskellFilePath
+        allDirPaths = scanl1 (</>) (splitPath dirPath)
```

Comment:
The function as written will `listDirectory` for all parent folders up to the root, then pick up the closest one. `listDirectory` could be extremely expensive, we should strive to do as little IO as possible. 

At the very least, let's reverse `allDirPaths` upfront instead of reversing `cabalFiles` after all `listDirectory` and `filterM` was done. But I would recommend to rewrite this function with explicit recursion, stripping components of `haskellFilePath` on each revursive call.

Reply:
Rewrote the function to an explicit recursion.